core/math/aabb.cpp

```diff
@@ -117,16 +117,15 @@ AABB AABB::intersection(const AABB &p_aabb) const {
 	return AABB(min, max - min);
 }
 
-bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *r_clip, Vector3 *r_normal) const {
+bool AABB::intersects_ray(const Vector3 &p_from, const Vector3 &p_dir, Vector3 *r_intersection_point, Vector3 *r_normal) const {
 #ifdef MATH_CHECKS
 	if (unlikely(size.x < 0 || size.y < 0 || size.z < 0)) {
 		ERR_PRINT("AABB size is negative, this is not supported. Use AABB.abs() to get an AABB with a positive size.");
 	}
 #endif
-	Vector3 c1, c2;
 	Vector3 end = position + size;
-	real_t near = -1e20;
-	real_t far = 1e20;
+	real_t tmin = 0.0;
```

Comment:
I'm short on time right now, but can you let us know why this figure should be changed to 0.0 (just in case this introduces float error)? A quick google suggested using -INFINITY, +INFINITY as before:

https://tavianator.com/2011/ray_box.html

Did you write this or copy from a source? It would be helpful to link to the source article in a comment, or at least provide reasoning for the change here.

Reply:
As the case where the component is equal to 0 is handled separately, I think this change makes sense.

| case 0 | case 1 | case 2 |
| :------: | :------: | :-------: |
| ![0](https://github.com/godotengine/godot/assets/30386067/07173a13-5bd3-4741-8211-621e2bb1e077) | ![1](https://github.com/godotengine/godot/assets/30386067/181b3059-9c94-4fbb-9f2c-db27d2304d4e) | ![2](https://github.com/godotengine/godot/assets/30386067/a016f354-37a4-4364-b014-becb1ee3c3f8) |

Less than 0 means far away from the point (pink line segment), and greater than 0 means close to the point (green line segment).

It is possible for the intersection point to fall on the green line segment, but it is impossible to fall on the pink line segment.

Reply:
Ah good explanation @Rindbee that makes sense. :+1: 

However this does brings up another issue, that calling code may in a lot of cases want to handle the start point inside the box as a special case. It would have been neater if the function returned not just whether it intersects, but also whether the start point is inside the box (being a special case), as this now might require a separate test.

Maybe we are stuck with this for backward compatibility.

Reply:
I'm not sure if **on** aabb box (vertex/edge/face)  is a special case (it seems to be considered **inside**), but it's special enough if the direction is parallel to some axis.

Reply:
It's changed to a zero because you don't want to negativly cast the ray. Worse case your ray origin is inside AABB and you stop. I've written these tests tons of times with more than 15 years in the game industry. The main book I use for ref is Real-Time Collision Detection by Christer Ericson if you need to cross check it.